Test setting of all config values; uniformly process config files

jdavcs opened the pull request:
[This is close to done.]

Fixes #9973 (here's the description of a specific bug: https://github.com/galaxyproject/galaxy/issues/9973#issuecomment-655792443)

**Overall problem.** We had tests (both integration and unit) for testing *default* values of configuration options. We did *not* test overriding those defaults. Which is why we were not catching at least one type of bugs (9973 shows one that surfaced). The issues are due to non-uniform processing of config options: we load them uniformly ([using defaults from the schema, adding some processing steps, etc.](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L114)). But then we reload some of them in `parse_config_file_options` - which overrides not only the schema defaults, but also any processing steps previously executed. User-set relative config file paths being resolved with respect to the root directory instead of their correct location is one example of a bug. Not being able to set any of the options that pass through `parse_config_file_options` to `None` is another unintended consequence (currently, the user *can* explicitly set all other config options to `None`; if we want to disallow this for these options, we can do that explicitly, and not as an implicit side-effect).

**What this PR contains:** 
- Tests for loading config option defaults (326 out of 329) + overriding defaults (302 out of 329). Tests for default values have been partially rewritten (they should be simpler now, much easier to read)
- Changes to config module, config schema + minor edits to a few other modules (those that made use of `app.config.foo_set`: this has been replaced with `app.config.is_set('foo')`)
- Minor edits to the testing infrastructure enabling testing of these options (mostly different handling of loading values into the mock config)
The config schema has been changed to make use of `path_resolves_to` instead of hardcoding prefixes (`config/`, `data/`, etc.).

The PR consists of many commits - that's intentional. Each commit is focused on one change; the commit message contains details about what was done and why. The "Refactor processing of ..." commit messages are not all the same.

This PR will be ready for final review in a few more commits: my goal is to remove `parse_config_file_options` and enable tests for the few remaining config options. Meanwhile, any feedback is, as always, very much appreciated.

When this is done, I'll write a "How-to" entry on configuration edits for the documentation.

jdavcs committed:
Refactor code for testing base config dirs

- Rearrange, simplify code
- Set fixture's scope to 'module' to reuse basic config across tests

jdavcs committed:
Enable testing of user-set values; refactor

- Do not hardcode schema defaults into path resolvers, so that we can
  test both schema defaults AND overwriting config properties
  with user-provided values
- Fix testing for sample_dir: add `.sample` suffix for expected value
- Pass both default and set-value to `get_expected_value()`: expected
  value is calculated based on key, default value AND set value.

jdavcs committed:
Enable testing for email_[allow/block]list_file

Patch to prevent GalaxyAppConfiguration from trying to load a nonexistant file

jdavcs committed:
Ensure data_dir path is absolute when set by user

Same as config_dir, managed_config_dir

jdavcs committed:
Uncomment/test fixed config options (+35)

jdavcs committed:
Add is_set(foo) method, to replace foo_set

Returns True if foo is set in galaxy.yml, including setting to None.

This will replace the [property]_set attribute used for a few config
file options.

Unit test included.

jdavcs committed:
Refactor processing of auth_config_file

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Run make config-rebuild (previous changes rebuilt)
- Replace call to appconfig.auth_config_file_set with
  appconfig.is_set('auth_config_file')
- Enable unit test

jdavcs committed:
Refactor processing of markdown_export_css

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Enable unit test

jdavcs committed:
Refactor processing of markdown_export_css_invocation_reports

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Enable unit test

jdavcs committed:
Refactor processing of markdown_export_css_pages

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Enable unit test

jdavcs committed:
Refactor processing of dependency_resolvers_config_file

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Run make config-rebuild
- Enable unit test

jdavcs committed:
Amend the behavior of config.is_set

If the user explicitly sets `foo` to its default value specified in the
schema, `config.is_set('foo')` should still return `True`.

jdavcs committed:
Refactor processing of job_config_file

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Run make config-rebuild
- Enable unit test

jdavcs committed:
Refactor processing of job_resource_params_file

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Run make config-rebuild
- Enable unit test

jdavcs committed:
Refactor processing of migrated_tools_config

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. managed_config_dir)
- Run make config-rebuild
- Enable unit test

jdavcs committed:
Refactor processing of object_store_config_file

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Run make config-rebuild
- Enable unit test

jdavcs committed:
Refactor processing of oidc_backends_config_file

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Run make config-rebuild
- Enable unit test
- Get rid of config.oidc_backends_config property; use schema property
  instead (oidc_backends_config_file)

jdavcs committed:
Refactor processing of oidc_config_file

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Run make config-rebuild
- Enable unit test
- Get rid of config.oidc_config property; use schema property
  instead (oidc_config_file)

jdavcs committed:
Enable integration tests for fixed config options(4)

jdavcs committed:
Refactor processing of shed_data_manager_config_file

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. managed_config_dir)
- Run make config-rebuild
- Replace call to appconfig.shed_data_manager_config_file_set with
  appconfig.is_set('shed_data_manager_config_file')
- Enable unit test

jdavcs committed:
Refactor processing of tool_sheds_config_file

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Run make config-rebuild
- Enable unit test

jdavcs committed:
Refactor processing of user_preferences_extra_conf_path

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Replace call to appconfig.user_preferences_extra_conf_path_set with
  appconfig.is_set('user_preferences_extra_conf_path')
- Run make config-rebuild
- Enable unit test

jdavcs committed:
Refactor processing of workflow_resource_params_file

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Run make config-rebuild
- Enable unit test

jdavcs committed:
Refactor processing of workflow_schedulers_config_file

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Run make config-rebuild
- Enable unit test

jdavcs committed:
Fix unit test for workflow_resource_params_mapper

jdavcs committed:
Refactor processing of file_sources_config_file

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Run make config-rebuild
- Enable unit test

jdavcs committed:
Refactor processing of shed_tool_config_file

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. managed_config_dir)
- Run make config-rebuild (previous changes rebuilt)
- Replace call to appconfig.shed_tool_config_file_set with
  appconfig.is_set('shed_tool_config_file')
- Enable unit test

jdavcs committed:
Refactor processing of shed_tool_data_table_config

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. managed_config_dir)
- Run make config-rebuild (previous changes rebuilt)
- Replace call to appconfig.shed_tool_data_table_config with
  appconfig.is_set('shed_tool_data_table_config')
- Modify test driver_utils: do not add option to mock config if default
  is None: this value will overwrite the schema default, even if set to
  None - thus causing an error when galaxy tries to parse the file
- Enable unit test

jdavcs committed:
Refactor processing of data_manager_config_file

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Add path_resolves_to to schema (resolves w.r.t. config_dir)
- Run make config-rebuild (previous changes rebuilt)
- Replace call to appconfig.data_manager_config_file_set with
  appconfig.is_set('data_manager_config_file')
- Modify test driver_utils: make the path to the sample config file
  absolute: resolving that path w.r.t to root in the config module was a
  bug and is fixed now. Thus, if a relative path is supplied, the config
  module will resolve it w.r.t config_dir - as per schema. Passing the
  file used for testing, therefore, must be an absolute path, since it
  is located under 'test/', not 'config/'.
- Enable unit test

jdavcs committed:
Refactor processing of file_path

Process like any other config option:
- Remove from list of defaults in parse_config_options()
- Replace call to appconfig.file_path_set with
  appconfig.is_set('file_path')
- Enable unit test

jdavcs committed:
Refactor to test allowed_origin_hostnames

- Enable testing of allowed_origin_hostnames config option
- Do not call kwargs.get() to obtain `allowed_origin_hostnames`: all
  kwargs are processed by BaseAppConfiguration uniformly.

jdavcs committed:
Fix test for job_config_file

jdavcs committed:
Fix test for data_manager_config_file

jdavcs committed:
Correct handling of is_set method in galaxy mock

Also, fix linting errors

jmchilton commented on test/integration/test_config_defaults.py:
```diff
@@ -161,7 +161,7 @@ def expected_default_config_dir(value):
     'config_dir',  # value overridden for testing
     'data_dir',  # value overridden for testing
-    'data_manager_config_file',  # broken: remove 'config/' prefix from schema
+    'data_manager_config_file',  # value overidden for testing
```
Can this whole test go? Are there comparable unit tests at this point?

Having to explicitly place everything we override in here is sort of gross and unexpected - also it clashes with GALAXY_CONFIG_OVERRIDE_ and such for ad hoc testing - which I guess could break all the integration tests, but yeah...

jdavcs replied:
Yes, it could go. If you think it doesn't serve its purpose or causes more problems than it solves, I don't mind dropping it (I'd still test the schema defaults though). That said, a few arguments:
1. It's for config options defined in the schema *only*. It's not for everything we override (or create) in the config module.
2. We load these options automatically (which is tested), but we do not test overriding them. The logic for overriding options varies, and in some cases we were doing it wrong (`parse_config_file _options`, as per #9973). This test should prevent such bugs.
3. At least in theory, this unit test should not affect any integration or other unit tests.

That said, I do agree it's ugly. I tried generalizing it and setting new values automatically based on data type, but there are  too many special cases - I couldn't come up with an elegant solution. So it's not elegant. If the test also feels unexpected, that's worse (I can't judge that myself - I'm too "used to it").

To sum up, if you still feel it's better dropped, I'll drop it. In that case, I think I'll write a few tests targeting specific individual options (like the various config files).

jdavcs replied:
I'll remove the tests generated for overriding schema defaults before opening this for review. Keeping them for now while I ensure all other tests pass (they do pass locally).

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -132,6 +133,14 @@ def _process_renamed_options(self, kwargs):
+        """Check if a configuration option has been explicitly set."""
+        # NOTE: This will check all supplied keyword arguments, including those not in the schema.
+        # To check only schema options, change the line below to `if property not in self._raw_config:`
+        if property not in self._kwargs:
```
```suggestion
        if key not in self._kwargs:
```

jdavcs replied:
@nsoranzo oh my! thank you for catching this!

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -132,6 +133,14 @@ def _process_renamed_options(self, kwargs):
+        # NOTE: This will check all supplied keyword arguments, including those not in the schema.
+        # To check only schema options, change the line below to `if property not in self._raw_config:`
+        if property not in self._kwargs:
+            log.warning("Configuration option does not exist: '%s'" % key)
```
Is this needed or is just for debug?

jdavcs replied:
Definitely not for debugging. As for needed or not, we had a discussion with @mvdbeek a while ago: 
https://github.com/galaxyproject/galaxy/pull/8867#pullrequestreview-306394120
My preference is to err on the side of caution, and I'd even raise an exception (see discussion^). However, if it safe to log a warning or even ignore this, I don't mind. @natefoo - `config.is_set('foo')`replaces `config.foo_set` - should we not guard/guard/guard aggressively against a case where `foo` is misspelled by a dev?

nsoranzo replied:
It could have made sense in #8867 to log this message when you were checking if the key is in `self._raw_config`, but now you are checking if it's in `self._kwargs` and it doesn't make sense to log if it's just not set by the admin.

jdavcs replied:
Another whoops. It should be checking `self._raw_config`, of course! Thank you. I've accidentally overwritten your previous fix (key > property), added it w/this fix.

jdavcs committed:
Fix bug in config.is_set()

jdavcs committed:
Patch tempfile.tempdir

From previous commit:
https://github.com/galaxyproject/galaxy/pull/9631#issuecomment-617008412
This line -
https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/__init__.py#L383
- sets tempfile.tempdir, which is a global variable that defines the
value of the default dir argument to the functions in Python's tempfile
module. Which breaks half of the tests, of course.

jdavcs committed:
Resolve path for trs_servers_config_file in schema

- Do not duplicate default file in config/__init__.py
- Update unit test
- Run make config-rebuild
